com.hazelcast.jet.core.CancellationTest.when_multipleClientsCancel_then_allSucceed

_3.0-maintenance_ b79453edce91c6ab3bfbf57d7574f89fdb572f01

Failed on jet-oss-maintenance-oracle-jdk11:

http://jenkins.hazelcast.com/view/Jet/job/jet-oss-maintenance-oracle-jdk11/47/testReport/junit/com.hazelcast.jet.core/CancellationTest/when_multipleClientsCancel_then_allSucceed/

Stacktrace:
```
Error Message
com.hazelcast.jet.core.JobNotFoundException: Job with id 02ec-e48f-5a80-0001 not found
Stacktrace
java.util.concurrent.ExecutionException: com.hazelcast.jet.core.JobNotFoundException: Job with id 02ec-e48f-5a80-0001 not found
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at com.hazelcast.jet.core.CancellationTest.when_multipleClientsCancel_then_allSucceed(CancellationTest.java:332)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.hazelcast.jet.core.JobNotFoundException: Job with id 02ec-e48f-5a80-0001 not found
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$terminateJob$8(JobCoordinationService.java:298)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$29(JobCoordinationService.java:878)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$30(JobCoordinationService.java:897)
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:67)
	at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:227)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:126)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:79)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:163)
	at com.hazelcast.jet.impl.AbstractJobProxy.terminate(AbstractJobProxy.java:158)
	at com.hazelcast.jet.impl.AbstractJobProxy.cancel(AbstractJobProxy.java:141)
	at com.hazelcast.jet.core.CancellationTest.lambda$when_multipleClientsCancel_then_allSucceed$3(CancellationTest.java:327)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	... 2 more
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java
@@ -278,6 +278,28 @@ public class JobCoordinationService {
         assertIsMaster("Cannot " + terminationMode + " job " + idToString(jobId) + " on non-master node");
 
         return submitToCoordinatorThread(() -> {
+            MasterContext masterContext = masterContexts.get(jobId);
+            if (masterContext != null) {
+                // User can cancel in any state, other terminations are allowed only when running.
+                // This is not technically required (we can request termination in any state),
+                // but this method is only called by the user. It would be weird for the client to
+                // request a restart if the job didn't start yet etc.
+                // Also, it would be weird to restart the job during STARTING: as soon as it will start,
+                // it will restart.
+                // In any case, it doesn't make sense to restart a suspended job.
+                JobStatus jobStatus = masterContext.jobStatus();
+                if (jobStatus != RUNNING && terminationMode != CANCEL_FORCEFUL) {
+                    throw new IllegalStateException("Cannot " + terminationMode + ", job status is " + jobStatus
+                            + ", should be " + RUNNING);
+                }
+
+                String terminationResult = masterContext.jobContext().requestTermination(terminationMode, false).f1();
+                if (terminationResult != null) {
+                    throw new IllegalStateException("Cannot " + terminationMode + ": " + terminationResult);
+                }
+                return;
+            }
+
             JobResult jobResult = jobRepository.getJobResult(jobId);
             if (jobResult != null) {
                 if (terminationMode == CANCEL_FORCEFUL) {
@@ -288,34 +310,14 @@ public class JobCoordinationService {
                         + " because it already has a result: " + jobResult);
             }
 
-            MasterContext masterContext = masterContexts.get(jobId);
-            if (masterContext == null) {
-                JobRecord jobRecord = jobRepository.getJobRecord(jobId);
-                String message = "No MasterContext found for job " + idToString(jobId) + " for " + terminationMode;
-                if (jobRecord != null) {
-                    // we'll eventually learn of the job through scanning of records or from a join operation
-                    throw new RetryableHazelcastException(message);
-                }
-                throw new JobNotFoundException(jobId);
+            JobRecord jobRecord = jobRepository.getJobRecord(jobId);
+            if (jobRecord != null) {
+                // we'll eventually learn of the job through scanning of records or from a join operation
+                throw new RetryableHazelcastException("No MasterContext found for job " + idToString(jobId) + " for "
+                        + terminationMode);
             }
 
-            // User can cancel in any state, other terminations are allowed only when running.
-            // This is not technically required (we can request termination in any state),
-            // but this method is only called from client. It would be weird for the client to
-            // request a restart if the job didn't start yet etc.
-            // Also, it would be weird to restart the job during STARTING: as soon as it will start,
-            // it will restart.
-            // In any case, it doesn't make sense to restart a suspended job.
-            JobStatus jobStatus = masterContext.jobStatus();
-            if (jobStatus != RUNNING && terminationMode != CANCEL_FORCEFUL) {
-                throw new IllegalStateException("Cannot " + terminationMode + ", job status is " + jobStatus
-                        + ", should be " + RUNNING);
-            }
-
-            String terminationResult = masterContext.jobContext().requestTermination(terminationMode, false).f1();
-            if (terminationResult != null) {
-                throw new IllegalStateException("Cannot " + terminationMode + ": " + terminationResult);
-            }
+            throw new JobNotFoundException(jobId);
         });
     }
 

